src/documents/models.py

```diff
@@ -314,6 +397,83 @@ def thumbnail_file(self):
     def thumbnail_url(self):
         return reverse("fetch", kwargs={"kind": "thumb", "pk": self.pk})
 
+    def set_filename(self, filename):
+        if os.path.isfile(self.filename_to_path(filename)):
+            self.filename = filename
+
+
+def delete_empty_directory(directory):
+    if len(os.listdir(directory)) == 0:
+        try:
+            os.rmdir(directory)
+        except os.error:
+            # Directory not empty
+            pass
+
+
+@receiver(models.signals.m2m_changed, sender=Document.tags.through)
+@receiver(models.signals.post_save, sender=Document)
+def update_filename(sender, instance, **kwargs):
+    if instance.filename is None:
```

Comment:
Why do we not want to set the filename if it is not populated yet?

Reply:
No, because the file might not have been saved yet (during consumption) and we will fail otherwise. Added documentation.